core/mount/mount_linux.go

```diff
@@ -251,12 +251,22 @@ func doPrepareIDMappedOverlay(lowerDirs []string, usernsFd int) (tmpLowerDirs []
 	}
 	cleanUp := func() {
 		for _, lowerDir := range tmpLowerDirs {
-			if err := unix.Unmount(lowerDir, 0); err != nil {
+			// Do a detached unmount so even if the resource is busy, the mount will be
+			// gone (eventually) and we can safely delete the directory too.
+			if err := unix.Unmount(lowerDir, unix.MNT_DETACH); err != nil {
```

Comment:
MNT_DETACH isn't a best option :p 
The cleanup is defer callback. It's hard to notify that caller to retry. 
Based on this fact, the change is acceptable to me. 

However, I think we can handle this issue in other way.
Based on `mount_setattr` manpage, the `MOUNT_ATTR_RDONLY` can focus the temporary mount in read-only. And no one can delete the content. 

> NOTE: Even if there are leaky mountpoint, containerd will clean it up during booting. https://github.com/containerd/containerd/blob/08037e7624e6e3c49e5f5448e75da6aa2b65aec7/cmd/containerd/command/main.go#L181

It's kind of enhancement and it needs to change `IdMapMount` interface.
It should be handled in the follow-up.

Reply:
@mbaynton are you interested in tackling this?

Reply:
Hmm, potentially. I'm most interested in investigating a new code path for Linux 6.13+ using the new file descriptor based layer setup and the new mount API. I think this will keep the idmap temp mounts off the host filesystem tree at all times, so the one and only accessor of them will be the desired overlayfs, so (I hypothesize) there will be no possibility of EBUSY failures / no need to use MNT_DETACH / no possibility of leaky mountpoints.

But an alternative implementation that removes `MNT_DETACH`, instead makes these read-only as another safeguard, and accepts the possibility that the unmount may fail without `MNT_DETACH` seems pretty easy to put together too.

Reply:
@rata it sounds like you already started a follow-up?

> In the follow-up PR this thing is removed and handled differently (IMHO, it is clearer).

I don't see one pushed to `rata/containerd` but are there improvements you've got?

Reply:
I should have something for this to address @fuweid's request in a week.

Reply:
https://github.com/containerd/containerd/pull/10955